src/runtime/http.rs

```diff
@@ -54,28 +69,47 @@ impl HttpClient {
         &self,
         method: Method,
         uri: &str,
-        headers: impl IntoIterator<Item = &'a (&'a str, &'a str)>,
-    ) -> reqwest::Result<String> {
-        let text = self.request(method, uri, headers).await?.text().await?;
+        headers: &'a [(&'a str, &'a str)],
+    ) -> Result<String, Box<dyn Error>> {
+        let res = self.request(method, Uri::from_str(uri)?, headers).await?;
 
+        let bytes = body::to_bytes(res.into_body()).await?;
+
+        let text = String::from_utf8(bytes.to_vec())?;
         Ok(text)
     }
 
-    /// Executes an HTTP equest and returns the response.
-    pub(crate) async fn request<'a>(
-        &self,
+    /// Executes an HTTP request and returns the response.
+    pub(crate) fn request<'a>(
+        &'a self,
         method: Method,
-        uri: &str,
-        headers: impl IntoIterator<Item = &'a (&'a str, &'a str)>,
-    ) -> reqwest::Result<Response> {
-        let response = headers
-            .into_iter()
-            .fold(self.inner.request(method, uri), |request, (k, v)| {
-                request.header(*k, *v)
-            })
-            .send()
-            .await?;
+        uri: Uri,
+        headers: &'a [(&'a str, &'a str)],
+    ) -> BoxFuture<'a, Result<Response<Body>, Box<dyn Error>>> {
+        Box::pin(async move {
+            let mut request = Request::builder().uri(&uri).method(&method);
+
+            for header in headers {
+                request = request.header(header.0, header.1);
+            }
+
+            let request = request.body(Body::empty())?;
+            let response = self.inner.request(request).await?;
+
+            if response.status().is_redirection() {
+                if let Some(Ok(location)) = response.headers().get(LOCATION).map(|u| u.to_str()) {
+                    if let (Some(scheme), Some(authority)) = (uri.scheme_str(), uri.authority()) {
+                        let uri = Uri::builder()
+                            .scheme(scheme)
+                            .authority(authority.as_str())
+                            .path_and_query(location)
+                            .build()?;
+                        return self.request(method, uri, headers).await;
```

Comment:
We should probably remove Authorization and other sensitive headers before following the redirect.

Reply:
I'm not sure what you mean here. This module is only used internally here (for AWS). As far as I see, none of the uses have any sensitive headers.